File: leetcode/hard/239-sliding/stack.py
"""
  @author Vortexx2

  Runtime - 1736 ms
  Memory Usage - 30.7 MB
"""
from typing import Deque, List
from collections import deque


class Solution:
  def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:

    if not len(nums) or k == 1 or len(nums) == 1:
      return nums

    res = []
    dq = deque([])

    for i in range(k):
      while (len(dq) and nums[dq[-1]] < nums[i]):
        dq.pop()

      dq.append(i)

    for i in range(k, len(nums)):

      res.append(nums[dq[0]])

      while (len(dq) and nums[dq[-1]] < nums[i]):
        dq.pop()

      while (len(dq) and dq[0] + k <= i):
        dq.popleft()

      dq.append(i)

    res.append(nums[dq[0]])

    return res


if __name__ == "__main__":

  sol = Solution()
  print(sol.maxSlidingWindow([1, 3, 1, 2, 0, 5], 3))

File: leetcode/hard/239-sliding/stack.cpp
/**
 * @file stack.cpp
 * @author Vortexx2
 * @brief 
 * @date 14-07-2021
 * 
 * Runtime - 248 ms
 * Memory Usage - 131.8 MB
 */
#include <iostream>
#include <vector>
#include <deque>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

class Solution
{
public:
  vector<int> maxSlidingWindow(vector<int> &nums, int k)
  {
    if (!nums.size() || nums.size() == 1 || k == 1)
    {
      return nums;
    }

    vector<int> res;
    deque<int> dq;

    for (int i = 0; i < k; i++)
    {
      while (!dq.empty() && nums[dq.back()] < nums[i])
      {
        dq.pop_back();
      }

      dq.push_back(i);
    }

    for (int i = k; i < nums.size(); i++)
    {
      res.push_back(nums[dq.front()]);

      while (!dq.empty() && nums[dq.back()] < nums[i])
      {
        dq.pop_back();
      }

      while (!dq.empty() && dq.front() + k <= i)
      {
        dq.pop_front();
      }

      dq.push_back(i);
    }

    res.push_back(nums[dq.front()]);

    return res;
  }
};